### PR TITLE
feat: dynamic batching for training and log prob stages

### PR DESCRIPTION
**Pull request by jiemingz**
# What does this PR do ?
image

# Issues
List issues that this PR closes ([syntax](https://docs.github.com/en/issues/tracking-your-work-with-issues/using-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword)):

# Usage
* **You can potentially add a usage example below**

```python
# Add a code snippet demonstrating how to use this 
```

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](/NVIDIA/reinforcer/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you run the unit tests and functional tests locally? Visit our [Testing Guide](/NVIDIA/reinforcer/blob/main/docs/testing.md) for how to run tests
- [ ] Did you add or update any necessary documentation? Visit our [Document Development Guide](/NVIDIA/reinforcer/blob/main/docs/documentation.md) for how to write, build and test the docs.

# Additional Information
* ...

### CODE REVIEW

File: nemo_reinforcer/models/policy/dtensor_policy_worker.py [hunk 1]
```diff
@@ -94,7 +95,13 @@ def get_cpu_state_dict(
     return new_state_dict
 
 
-@ray.remote
+@ray.remote(
+    runtime_env={
+        "env_vars": {
+            "PYTORCH_CUDA_ALLOC_CONF": "expandable_segments:True"
```
Comment (terrykong) [inline nemo_reinforcer/models/policy/dtensor_policy_worker.py]: was there a new OOM?

File: nemo_reinforcer/distributed/batched_data_dict.py [hunk 2]
```diff
@@ -253,15 +332,22 @@ def repeat_interleave(self, num_repeats: int) -> "BatchedDataDict":
...
+        microbatch_indices = None,
     ) -> Iterator["SlicedDataDict"]:
         """Make an iterator over the batch that yields microbatches of size microbatch_size."""
-        bsize = self.size
-        assert bsize % microbatch_size == 0, (
-            f"Data dict size ({bsize}) is not a multiple of the provided microbatch size ({microbatch_size})"
-        )
-        for i in range(0, bsize, microbatch_size):
-            yield self.slice(i, i + microbatch_size)
+        if microbatch_indices is not None:
```
Comment (terrykong) [inline nemo_reinforcer/distributed/batched_data_dict.py]: prob should assert xor

File: nemo_reinforcer/models/policy/dtensor_policy_worker.py [hunk 3]
```diff
@@ -239,6 +247,24 @@ def get_gpu_info(self):
         """Return information about the GPU being used by this worker."""
         return get_gpu_info(self.model)
 
+
+    def get_microbatch_iterator(
```
Comment (terrykong) [inline nemo_reinforcer/models/policy/dtensor_policy_worker.py]: is this used anywhere?

File: nemo_reinforcer/models/policy/dtensor_policy_worker.py [hunk 4]
```diff
@@ -380,19 +432,30 @@ def get_logprobs(self, data: BatchedDataDict) -> BatchedDataDict:
...
+                    microbatch_indices=mbs_indices)
+            else:
+                mb_iterator = local_global_batch.make_microbatch_iterator(
+                    microbatch_size=logprob_batch_size)
 
-                # Create attention mask
+            for lp_batch in mb_iterator:
+                input_ids = lp_batch.get("input_ids").cuda()
                 input_lengths = lp_batch.get("input_lengths")
+                padded_seqlen = ((max(input_lengths) + 64 - 1) // 64) * 64
```
Comment (terrykong) [inline nemo_reinforcer/models/policy/dtensor_policy_worker.py]: could you add a comment on why 64? also, would we have an issue if the data was padded to a multiple of TP(b/c SP enabled) via `make_sequence_length_divisible_by`, and then this padding is not a multiple of the TP?

File: nemo_reinforcer/models/policy/dtensor_policy_worker.py (hunk 1 above)
Comment (jiemingz) [reply nemo_reinforcer/models/policy/dtensor_policy_worker.py]: the dynamic batch and sequence dimensions did cause memory usage increase due to memory allocator fragmentation, about 5gb. I feel like theres no harm in enabling this unless there's any objection?

Comment (terrykong) [reply nemo_reinforcer/models/policy/dtensor_policy_worker.py]: i think it's okay. maybe you can include a comment close to this declaration.

one thing, i'm wondering about is if this is even being used.

https://github.com/NVIDIA/nemo-rl/blob/506910a14eed03663f632347409f5505e5af09f6/nemo_rl/distributed/worker_groups.py#L171

I'm not sure if the env var set in the initial ray.remote call site is "merged" or "overridden". might be good to confirm

File: examples/configs/grpo_math_1B.yaml [hunk 5]
```diff
@@ -48,7 +48,13 @@ policy:
     sequence_parallel: false
     activation_checkpointing: false
     tensor_parallel_size: 1
-  
+
+  dynamic_batching:
```
Comment (parthchadha) [inline examples/configs/grpo_math_1B.yaml:59]: Lets add a small comment what `dynamic_batching` is and when should it be enabled? Also, explain the args too?

File: nemo_rl/distributed/batched_data_dict.py [hunk 6]
```diff
@@ -141,6 +149,7 @@ def shard_by_batch_size(
         shards: int,
         batch_size: Optional[int] = None,
         allow_uneven_shards: bool = False,
+        dynamic_batching_cfg = None,
```
Comment (parthchadha) [inline nemo_rl/distributed/batched_data_dict.py]: Lets make this typed dict?

File: nemo_rl/distributed/batched_data_dict.py [hunk 7]
```diff
@@ -218,13 +234,62 @@ def shard_by_batch_size(
...
+                else:
+                    sorted_v = [v[i] for i in unsorted_indices]
+                self.data[k] = sorted_v
+
+        # if using dynamic microbatching, preprocess the data by sorting the data 
+        # by the sequence lengths. This ensures each DP rank receives samples of about 
+        # equal sequence lengths which improves load balancing
+        if dynamic_batching_cfg is not None:
+            batch_sorted_indices = []
+            for chunk_idx in range(num_chunks):
```
Comment (parthchadha) [inline nemo_rl/distributed/batched_data_dict.py:306]: It will be great if you can add some unit tests specifically for sorting, chunking added in this PR.

File: nemo_rl/distributed/batched_data_dict.py [hunk 8]
```diff
@@ -26,7 +26,25 @@
...
+    input_lengths_key: str
+
 class BatchedDataDict(UserDict, Generic[DictT]):
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.metadata = {
+            "is_sorted" : False,
+            "micro_batch_indices": None
+        }
```
Comment (terrykong) [inline nemo_rl/distributed/batched_data_dict.py]: any chance we can have these top level? to make discoverability easier?

```suggestion
        self.is_sorted = False
        self.micro_batch_indices = None
        self.sort_indices = None
```
